src/realm/sync/client_syncsocket.hpp

```diff
@@ -0,0 +1,235 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#ifndef REALM_SYNC_CLIENT_SYNC_SOCKET_HPP
+#define REALM_SYNC_CLIENT_SYNC_SOCKET_HPP
+
+#include <map>
+#include <memory>
+#include <string>
+
+#include <realm/status.hpp>
+
+#include <realm/sync/config.hpp>
+
+#include <realm/util/functional.hpp>
+#include <realm/util/optional.hpp>
+
+namespace realm::sync {
+
+/// Struct that defines the endpoint to create a new websocket connection.
+/// Many of these values come from the SyncClientConfig passed to SyncManager when
+/// it was created.
+struct Endpoint {
+    using port_type = sync::port_type;
+
+    std::string address;   // Host address
+    port_type port;        // Host port number
+    std::string path;      // Includes access token in query.
+    std::string protocols; // One or more websocket protocols, separated with ", "
+    bool is_ssl;           // true if SSL should be used
+
+    /// The remaining fields are just pass through values from SyncConfig. They may
+    /// not be provided if the SDK chooses not to support the related config options.
+    /// These may be necessary when using websocket libraries without low-level
+    /// control. These items may be removed or converted to test only parameters
+    /// in the future.
+    std::map<std::string, std::string> headers; // Only includes "custom" headers.
+    bool verify_servers_ssl_certificate;        // If true, verify server ssl certs
+    util::Optional<std::string> ssl_trust_certificate_path;
+    std::function<SyncConfig::SSLVerifyCallback> ssl_verify_callback;
+    util::Optional<SyncConfig::ProxyConfig> proxy; // Send traffic through a proxy
+};
+
+
+/// The WebSocket base class that is used by the SyncClient to send data over the
+/// WebSocket connection with the server. This is the class that is returned by
+/// SyncSocketInterface::connect() when the connection to an endpoint is requested.
+/// If an error occurs while establishing the connection, the error is presented
+/// to the WebSocketObserver provided when the WebSocket was created.
+struct WebSocketInterface {
+    /// The destructor must close the websocket connection when the WebSocket object
+    /// is destroyed
+    virtual ~WebSocketInterface() = default;
+
+    /// Write data asynchronously to the WebSocket connection. The handler function
+    /// will be called when the data has been sent successfully. The WebSocketOberver
+    /// provided when the WebSocket was created will be called if any errors occur
+    /// during the write operation.
+    virtual void async_write_binary(const char* data, size_t size, util::UniqueFunction<void()>&& handler) = 0;
```

Comment:
Should this take a `util::Span<const char>` for data/size?

Reply:
Sure - I can update it